### DISCUSSION TITLE
How do I put color same?

### DISCUSSION THREAD

**Discussion by tanujdhiman** [ultralytics/yolov5#4339] [Q&A]
Hey everybody !!

I am trying this model but is there any feature in this model which provides you same color of bounding box and object detected.

Thanks

**Comment by IneovaAI** [comment]
Hi, to clarify, do you want to make all the bounding boxes the same color or do you want to make them the same color as the object?

**Reply by tanujdhiman** [reply]
I want to make same color of bounding box and object. 
For example ::- Suppose Detector is detecting a red hat so color of bounding box is also red. 

Hope you are understanding. 
Thanks

**Reply by IneovaAI** [reply]
I do, however, I don't know if such a function is already built-in, I doubt it. I will give it a go and report back if I crack it.

**Reply by IneovaAI** [reply]
![apple-and-banana-stock-image-959840](https://user-images.githubusercontent.com/67843470/128643178-1c793a23-d1ca-40d7-ad71-56e026246a76.jpg)
![depositphotos_135793274-stock-photo-orange-apple-and-banana](https://user-images.githubusercontent.com/67843470/128643179-80554071-c316-40fc-ab60-536b885d2d90.jpg)
@tanujdhiman Do you want something like this?

**Reply by tanujdhiman** [reply]
Yes exactly.

**Reply by IneovaAI** [reply]
In plots.py change the def plot_one_box with this:

def plot_one_box(x, im, raw_img, color=(128, 128, 128), label=None, line_thickness=3):
    # We bring in the raw image to do better average on colors
    # Plots one bounding box on image 'im' using OpenCV
    detect_color = True
    assert im.data.contiguous, 'Image not contiguous. Apply np.ascontiguousarray(im) to plot_on_box() input image.'
    tl = line_thickness or round(0.002 * (im.shape[0] + im.shape[1]) / 2) + 1  # line/font thickness
    c1, c2 = (int(x[0]), int(x[1])), (int(x[2]), int(x[3]))

    if label:
        tf = max(tl - 1, 1)  # font thickness
        t_size = cv2.getTextSize(label, 0, fontScale=tl / 3, thickness=tf)[0]
        c3 = c1[0] + t_size[0], c1[1] - t_size[1] - 3  # This get the top right coordinate of the label box

        if detect_color:
            xywh = [c1[0], c1[1], c2[0] - c1[0], c2[1] - c1[1]]  # Build the detect box xy top left, width and height
            xywh = [0 if seg < 0 else seg for seg in xywh]  # Change the coordinates of all boxes to be inside the image.

            # First we crop the box from the image
            sub_img = raw_img[xywh[1]:xywh[1] + xywh[3], xywh[0]:xywh[0] + xywh[2]]

            # Then we get the dominant color in that box
            # This is an answer on stackoverflow by Tonechas
            # link https://stackoverflow.com/questions/43111029/how-to-find-the-average-colour-of-an-image-in-python-with-opencv
            pixels = np.float32(sub_img.reshape(-1, 3))
            n_colors = 1
            criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 200, .1)
            flags = cv2.KMEANS_RANDOM_CENTERS
            _, labels, palette = cv2.kmeans(pixels, n_colors, None, criteria, 10, flags)
            _, counts = np.unique(labels, return_counts=True)
            dominant = palette[np.argmax(counts)].astype('float64')

            # Here we brighten the color
            scaler = 255/np.max(dominant)
            dominant =dominant * scaler

            color = dominant

        # Here we decide if the text must be black or white, if the label color is light then text is dark
        # and vice versa.
        text_color = [255 if np.average(color) < 255 / 2 else 0]
        text_color = np.zeros(3) + text_color

        # We then plot it all on the im
        cv2.rectangle(im, c1, c2, color, thickness=tl, lineType=cv2.LINE_AA)  # Object box
        cv2.rectangle(im, c1, c3, color, -1, cv2.LINE_AA)  # Label Box
        cv2.putText(im, label, (c1[0], c1[1] - 2), 0, tl / 3, text_color, thickness=tf, lineType=cv2.LINE_AA)

**Reply by IneovaAI** [reply]
Also add:
raw_img = im0s.copy()
into line 126 in detect.py above process predictions so it looks like this:

 # Second-stage classifier (optional)
        if classify:
            pred = apply_classifier(pred, modelc, img, im0s)

        raw_img = im0s.copy()  # <<LINE 126 is HERE!
        # Process predictions
        for i, det in enumerate(pred):  # detections per image
            if webcam:  # batch_size >= 1
                p, s, im0, frame = path[i], f'{i}: ', im0s[i].copy(), dataset.count

**Reply by IneovaAI** [reply]
Then when calling the function add the raw_image:

                    if save_img or save_crop or view_img:  # Add bbox to image
                        c = int(cls)  # integer class
                        label = None if hide_labels else (names[c] if hide_conf else f'{names[c]} {conf:.2f}')
                        plot_one_box(xyxy, im0, raw_img=raw_img, label=label, color=colors(c, True), line_thickness=line_thickness)
This is on line 160 in detect.py for me ^^^^^^^^^^^^^ look here
